spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/HystrixTargeter.java

```diff
@@ -29,26 +32,65 @@
+		Class<?> fallback = factory.getFallback();
+		if (fallback != void.class) {
+			return targetWithFallback(factory.getName(), context, target, builder, fallback);
+		}
+		Class<?> fallbackFactory = factory.getFallbackFactory();
+		if (fallbackFactory != void.class) {
+			return targetWithFallbackFactory(factory.getName(), context, target, builder, fallbackFactory);
+		}
+
+		return feign.target(target);
+	}
+
+	private <T> T targetWithFallbackFactory(String feignClientName, FeignContext context,
+											Target.HardCodedTarget<T> target,
+											HystrixFeign.Builder builder,
+											Class<?> fallbackFactoryClass) {
+		FallbackFactory<? extends T> fallbackFactory = (FallbackFactory<? extends T>)
+			getFromContext(feignClientName, context, fallbackFactoryClass, FallbackFactory.class);
+		Object exampleFallback = fallbackFactory.create(new RuntimeException());
+		Assert.notNull(exampleFallback,
+			String.format(
+			"Incompatible fallbackFactory instance for feign client %s. Factory may not produce null!",
+				feignClientName));
+		if (!target.type().isAssignableFrom(exampleFallback.getClass())) {
+			throw new IllegalStateException(
+				String.format(
+					"Incompatible fallbackFactory instance for feign client %s. Factory produces instances of '%s', but should produce instances of '%s'",
+					feignClientName, exampleFallback.getClass(), target.type()));
+		}
+		return builder.target(target, fallbackFactory);
+	}
 
-		Object fallbackInstance = context.getInstance(factory.getName(), factory.getFallback());
+
+	private <T> T targetWithFallback(String feignClientName, FeignContext context,
+									 Target.HardCodedTarget<T> target,
+									 HystrixFeign.Builder builder, Class<?> fallback) {
+		T fallbackInstance = getFromContext(feignClientName, context, fallback, target.type());
+		return builder.target(target, fallbackInstance);
+	}
+
+	private <T> T getFromContext(String feignClientName, FeignContext context,
+								 Class<?> annotationType, Class<T> targetType) {
+		Object fallbackInstance = context.getInstance(feignClientName, annotationType);
 		if (fallbackInstance == null) {
 			throw new IllegalStateException(String.format(
-					"No fallback instance of type %s found for feign client %s",
-					factory.getFallback(), factory.getName()));
+					"No fallback/fallbackFactory instance of type %s found for feign client %s",
+				annotationType, feignClientName));
 		}
 
-		if (!target.type().isAssignableFrom(factory.getFallback())) {
+		if (!targetType.isAssignableFrom(annotationType)) {
 			throw new IllegalStateException(
 					String.format(
-							"Incompatible fallback instance. Fallback of type %s is not assignable to %s for feign client %s",
-							factory.getFallback(), target.type(), factory.getName()));
+							"Incompatible fallback/fallbackFactory instance. Fallback/fallbackFactory of type %s is not assignable to %s for feign client %s",
```

Comment:
I'd rather have a specific message. Pass 'fallback' or 'fallbackFactory' into the method.